src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/OverlayPane.java

```diff
@@ -0,0 +1,201 @@
+import java.awt.Component;
+import java.awt.Container;
+import java.awt.Dimension;
+import java.awt.Graphics;
+import java.awt.Graphics2D;
+import java.awt.Point;
+import java.awt.Rectangle;
+import java.awt.Window;
+import java.awt.geom.Area;
+import java.util.HashSet;
+import java.util.Set;
+import javax.swing.JFrame;
+import javax.swing.JOptionPane;
+import javax.swing.JPanel;
+import javax.swing.JRootPane;
+import javax.swing.SwingUtilities;
+import org.jetbrains.annotations.NotNull;
+
+public class OverlayPane extends JPanel {
+  private static OverlayPane activeOverlay = null;
+
+  private final JRootPane associatedRootPane;
+  private final Set<Component> exemptComponents;
+  private final Set<BalloonPopup> popups;
+
+  private void revalidatePane() {
+    getRootPane().revalidate();
+    getRootPane().repaint();
+  }
+
+  @Override
+  protected void paintComponent(Graphics graphics) {
+    super.paintComponent(graphics);
+
+    for (var c : popups) {
+      c.recalculateBounds();
+    }
+
+    // use Graphics's copy so as not to influence subsequent calls using the same Graphics object
+    Graphics2D g = (Graphics2D) graphics.create();
+    g.setComposite(AlphaComposite.SrcOver.derive(0.7f));
+    g.setColor(Color.BLACK); // using JBColor.BLACK is wrong here
+
+    var overlayArea = new Area(new Rectangle(0, 0, getWidth(), getHeight()));
+
+    for (var c : exemptComponents) {
+      // convertPoint is necessary because the component uses a different coordinate origin
+      var windowPos = SwingUtilities.convertPoint(c, c.getX(), c.getY(), this);
+      var componentRect = new Rectangle(windowPos.x, windowPos.y, c.getWidth(), c.getHeight());
+
+      overlayArea.subtract(new Area(componentRect));
+    }
+
+    for (var c : popups) {
+      // popups are already places in the overlay's coordinate system
+      var componentRect = new Rectangle(c.getX(), c.getY(), c.getWidth(), c.getHeight());
+      overlayArea.subtract(new Area(componentRect));
+    }
+
+    g.fill(overlayArea);
+  }
+
+  @Override
+  public JRootPane getRootPane() {
+    return associatedRootPane;
+  }
+
+  @Override
+  public int getWidth() {
+    return associatedRootPane.getLayeredPane().getWidth();
+  }
+
+  @Override
+  public int getHeight() {
+    return associatedRootPane.getLayeredPane().getHeight();
+  }
+
+  @Override
+  public Dimension getSize() {
+    return new Dimension(getWidth(), getHeight());
+  }
+
+  @Override
+  public Dimension getPreferredSize() {
+    return getSize();
+  }
+
+  @Override
+  public Dimension getMinimumSize() {
+    return getSize();
+  }
+
+  @Override
+  public Dimension getMaximumSize() {
+    return getSize();
+  }
+
+  @Override
+  public boolean isOpaque() {
+    return false;
+  }
+
+  @Override
+  public boolean isVisible() {
+    return true;
+  }
+
+  public static boolean isOverlayInstalled() {
+    return activeOverlay != null;
+  }
+
+  /**
+   * Installs the overlay.
+   */
+  public static void installOverlay() {
```

Comment:
Can only be called from EDT, i assume? That's fine but it's best to be explicitly stated.